src/marzullo.zig

```diff
@@ -0,0 +1,306 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+/// Marzullo's algorithm, invented by Keith Marzullo for his Ph.D. dissertation in 1984, is an
+/// agreement algorithm used to select sources for estimating accurate time from a number of noisy
+/// time sources. NTP uses a modified form of this called the Intersection algorithm, which returns
+/// a larger interval for further statistical sampling. However, here we want the smallest interval.
+pub const Marzullo = struct {
+
+    /// The smallest interval consistent with the largest number of sources.
+    pub const Interval = struct {
+        /// The lower bound on the minimum clock offset.
+        lower_bound: i64,
+
+        /// The upper bound on the maximum clock offset.
+        upper_bound: i64,
+
+        /// The number of "true chimers" consistent with the largest number of sources.
+        sources_true: u8,
+
+        /// The number of "false chimers" falling outside this interval.
+        /// Where `sources_false` plus `sources_true` always equals the total number of sources.
+        sources_false: u8,
+    };
+
+    /// A tuple represents either the lower or upper end of a bound, and is fed as input to the
+    /// Marzullo algorithm to compute the smallest interval across all tuples.
+    /// For example, given a clock offset to a remote replica of 3s, a round trip time of 1s, and
+    /// a maximum tolerance between clocks of 100ms on either side, we might create two tuples, the
+    /// lower bound having an offset of 2.4s and the upper bound having an offset of 3.6s,
+    /// to represent the error introduced by the round trip time and by the clocks themselves.
+    pub const Tuple = struct {
+        source: u8,
```

Comment:
is `source` an identifier?

Reply:
Yes, it's the index of the clock source in the `sources` array kept by the layer above this (not visible here). It's the same as the  `replica` index in `replica.zig` but I used the terminology from the algorithm instead of calling it `replica`. I'll add a comment to clarify what this is.